### ISSUE TITLE
support registration of 3bots

### ISSUE THREAD

**Issue by GlenDC** [threefoldfoundation/tfchain#189]
The feature request is tracked in rivine/home/#9 and specced in <https://github.com/rivine/home/blob/master/specs/3bot_registration.md>.

A new transaction type (`0x90`), will allow the registration of 3bots. The transaction would be encoded in JSON as follows:

```javascript
{
    "version": 144, // 0x90
    "data": {
        "address": <unlockhash>,
        // each dns is a string in a defined format
        "dns": { // link one or multiple dns names to one of multiple network addresses
            "char5": [networkAddressA, networkAddressB, ...], // dns name format to be defined
            "char5.char5": [networkNameA, networkAddressC, ...],
        },
        "validuntil": <epoch>, // -1 could be reserved to allow the disabling of this property
        // ed25519 public key
        "publickey": "ed25519:28c1edd4c35f662cccfa7fc02194959d75855c02d342c1131b110c9e96764d9b",
        "minerfees": [],
        // ed25519 signature, using all other properties as input
        "signature": "433f5283a82fac28dacddeb98ed57aabb649a4aad2e7813af8a0009e0d035f625724dc7ef9ef39e75aef10fc77c4ed43fa0ce09f80c77d81ffd0c04ee7ca8c00"
    }
}
```

Open Questions:

* is the bot's TFT address a regular address (`0x01`/`0x03`), or do we want to create a new type of unlock address just for bots? If so how to differentiate them from regular users? Could for example be `0x10` with the hash based on the public key;
* the active bool has been removed (see <https://github.com/rivine/home/issues/9#issuecomment-421422404> why), if this is not OK, let me know;
* for now I did the dns names and ips (proposing to support any type of network address by the way, which also includes ipv4 and ipv6) as a mapping, rather than separate listing, this allows the different dns names to point to different addresses, if you want the ability to have multiple dns names to the same addresses without having to redefine all those addresses each time, we could also have the `"dns"` property as `"dns":[{"names":["aaaaa.bbbbb", "aaaaa.ccccc"],"addresses":["a1", "a2","a3"]},{"names":["aaaaa.ddddd"],"addresses":["a4"]}]` instead;
* do we need any special validation on the validity period? e.g. a a maximum time window, or can it also be possible to allow a bot to be forever valid? Doesn't seem like a good idea, but it's a decision to be made;
* is there a maximum amount of DNS entries we want to allow per public key?
* Is there a maximum amount of network address in total and/or per DNS entry?
* In the spec it is implicitly defined that public keys of a 3bot can be changed. If so it would mean we cannot derive the 3bot tft address from the public key, but perhaps that is OK?

Consensus rules for the 3bot registration Tx (`0x90`):

* The signature needs to match the transaction data;
* The TFT address has to be valid;
* The DNS names need be unique (and thus available at the point of registration);
* The DNS names need to be valid according to the given format:
  * a dns name can have multiple parts, each separated by a dot (`.`);
  * each part needs to have at least 5 characters;
  * for each part the following character range is allowed: `[a-zA-Z0-9]` (but a part cannot be made of numeric characters only);
  * the maximum length of a DNS name is 63 characters (including the separator `.` character and all parts of that name);
* Each address has to be a valid domain name or IPv4 address or IPv6 address;
* Each DNS entry requires at least one network address;

The registration-specific fees and consensus rules around it have been moved to a separate issue. See issue #190 for this.

The Unique ID is not part of the registration Tx, given the unique ID is the public key. the Unique ID per bot is incremental and given by any explorer/dns node, rather than as part of the Tx. When a new public key appears it receives the next sequential unique ID (a `64 bit` integer). A registration of an earlier block has therefore a lower unique ID than that of a block later. When doing a registration using a public key that was already known, there will not be a new unique ID assigned, and instead the already existing record will be updated. The unique ID maps to the public key, validity period and DNS entries. Using the public key, one can request the unique ID of the bot as defined with its record.

> Personally I would remove the entire incremental unique ID completely, and always use the public key (or tft address derived from that public key) as the unique ID. it is much more natural, and anyhow, the bot will have to use its public key (or tft address derived from that public key) in order to get its unique ID the first time, given that sequential ID would only be possible to be assigned once the registration is part of a created block. If there is a however a good reason to have this in there, we can see how to make this work in the best way possible. But assigning it while creating the tx is not possible, given the order of that tx within the context of the chain is not yet known at that time, and for now the block creators have no way to add extra info to a transaction, nor should they.

**Comment by robvanmieghem** [comment]
Open Questions:

* is the bot's TFT address a regular address (`0x01`/`0x03`), or do we want to create a new type of unlock address just for bots? If so how to differentiate them from regular users? Could for example be `0x10` with the hash based on the public key;
**I would use a regular address**
* the active bool has been removed (see <https://github.com/rivine/home/issues/9#issuecomment-421422404> why), if this is not OK, let me know;
**OK**
* for now I did the dns names and ips (proposing to support any type of network address by the way, which also includes ipv4 and ipv6) as a mapping, rather than separate listing, this allows the different dns names to point to different addresses, if you want the ability to have multiple dns names to the same addresses without having to redefine all those addresses each time, we could also have the `"dns"` property as `"dns":[{"names":["aaaaa.bbbbb", "aaaaa.ccccc"],"addresses":["a1", "a2","a3"]},{"names":["aaaaa.ddddd"],"addresses":["a4"]}]` instead;
**fine for now**
* do we need any special validation on the validity period? e.g. a a maximum time window, or can it also be possible to allow a bot to be forever valid? Doesn't seem like a good idea, but it's a decision to be made;
* is there a maximum amount of DNS entries we want to allow per public key?
**start with 10**
* Is there a maximum amount of network address in total and/or per DNS entry?
**3 per dns**

Consensus rules for the 3bot registration Tx (`0x90`):

* Each DNS entry requires at least one network address;
**I would not do this, names can be reserved before actual deployment then**


**Looks good for the rest**

**Comment by GlenDC** [comment]
> do we need any special validation on the validity period? e.g. a a maximum time window, or can it also be possible to allow a bot to be forever valid? Doesn't seem like a good idea, but it's a decision to be made;

You didn't answer this one @robvanmieghem.

---

Also shouldn't the DNS registration fee be PER dns name->address link? Or at least have it be influenced by it.

**Comment by GlenDC** [comment]
@robvanmieghem here is my latest proposal for how the 3bot registration Tx could look like:

```javascript
{
    "version": 144, // 0x90
    "data": {
        // public address of the wallet of the bot,
        // only `0x01` and `0x03` addresses are allowed,
        // that can be used by others to put money in the bot's wallet
        "wallet": <unlockhash>,

        "net": [ // at least one network address is required, max 30
            // can be a domain name
            "network.address.a.com",
            "network.address.b.com",
            // can be an IPv4 address
            "83.200.201.201", ""],
            // can also be an IPv6 address
            "2001:db8:85a3::8a2e:370:7334"
        ],
        "dns": { // optional dns records, maximum 10
            // a DNS record links to one or multiple addresses
            // using its positional index as defined in the netaddresses property,
            // for example 1 would refer here to "network.address.b.com"
            "char5": [0],
            // a DNS record can have multiple network addresses linked to it,
            // max 3 network addresses per DNS record can be used
            "char5.char5": [0, 2],\
            // leave empty to just reserve the DNS name
            "aaaaa": [],
        }

        // usually an epoch unix time stamp in seconds,
        // but has some special values:
        // `0`:
        //   is reserved to mark this bot as inactive, and thus disable it,
        //   DNS entries will be reserved until the remaining paid time runs out,
        //   but the entries won't point to anything any longer
        // any other value has to be higher than the timestamp of the block is part of.
        "validuntil": <epoch>,

        // the regular Tx fee has to be paid as well,
        // see https://github.com/threefoldfoundation/tfchain/issues/190 for additional fees
        "fees": ["100000000"],

        // coin inputs used to fund the fees (same as regular Txs)
        "coininputs": [],
        // optional, and is only allowed to be used for refunds
        "coinoutputs": [],

        // public key (ed25519 by default, and only algo supported for now)
        "publickey": "ed25519:28c1edd4c35f662cccfa7fc02194959d75855c02d342c1131b110c9e96764d9b",

        // signature (verifiable using the given public key),
        // (ed25519 by default, and only algo supported for now)
        "signature": "433f5283a82fac28dacddeb98ed57aabb649a4aad2e7813af8a0009e0d035f625724dc7ef9ef39e75aef10fc77c4ed43fa0ce09f80c77d81ffd0c04ee7ca8c00"
    }
}
```

Notes:

* In order to change a public key of a 3bot I would use a separate Tx version for that, if that feature really has to be supported, because I don't really see how to fit it in nicely with this Tx. Unless we define some optional properties such as `origpublickey` and `origsignature`, that could work as well, but feels weird;
* I think if we force each 3bot registration Tx to define the full list of addresses and DNS entries, it would make it much cleaner and easier to follow. If we do want to allow partial updates, than my proposal to let the dns records refer directly to the sequential index of the network addresses won't really work, because how would you ever be able to delete a network address or replace it?

Extra Notes:
* I would also still use a separate wallet address for bots `0x05`, as to make it clear they are not humans. The rest would be the same though, you can send coins to it and whatever, and the public key will be the same as the used public key as well. In that case we do not even need the `wallet` property, as that can be derived from the public key of the bot;

Fees are ignored for sure, is in a separate issue #190 for a reason. Let's also please align there @robvanmieghem when you can.

How we are going to do the indexing and whether or not we want to use a separate (sequential?) unique ID per 3bot, is still the question, and something which is still undefined for now.

**Comment by GlenDC** [comment]
Personally I would drop the support for public key change, as the public key is what identifies a 3bot. I don't really see the need for it, regular wallets don't really have such support either. At the same time I do see that one might want to be able to transfer DNS names from one public key to another, but I think we can allow the definition of a 3bot multisig Tx for that, where bot 3bots have to agree, and where the DNS receiving 3bot pays for the DNS registration (will be free for the sending 3bot).

**Comment by GlenDC** [comment]
An initial stab at such DNS transfer Tx would be:

```javascript
{
    "version": 145, // 0x91
    "data": {
        sender: {
            // public key (ed25519 by default, and only algo supported for now)
            "publickey": <publickey_sender>,
            // signature (verifiable using the given public key),
            // (ed25519 by default, and only algo supported for now)
            // (!) Signature includes the public key of both parties as input,
            //     but the signatures not, so when signing, the initial party,
            //     has to make sure to already define the public key of the other
            //     party as well
            "signature": <signature_sender>,
        },
        receiver: {
            // public key (ed25519 by default, and only algo supported for now)
            "publickey": <publickey_receiver>,
            // signature (verifiable using the given public key),
            // (ed25519 by default, and only algo supported for now)
            // (!) Signature includes the public key of both parties as input,
            //     but the signatures not, so when signing, the initial party,
            //     has to make sure to already define the public key of the other
            //     party as well
            "signature": <signature_receiver>,
        },
        
        // public address of the wallet of the bot,
        // only `0x01` and `0x03` addresses are allowed,
        // that can be used by others to put money in the bot's wallet
        "wallet": <unlockhash>,

        // TODO: work out how the DNS transfer would look like,
        // in the most simple approach it would look like:
        //   "dns": ["dnsA", "dnsB", ...], // list of DNS entries to transfer
        // but that would mean the receiving 3bot would have to pay
        // for both the reservation of that DNS, and than for the actual usage for it,
        // while it would be nice if the receiver can already define addresses for it,
        // but given the current Tx 0x90 proposal, I do not see how that would work

        // the regular Tx fee has to be paid as well,
        // see https://github.com/threefoldfoundation/tfchain/issues/190 for additional fees
        "fees": ["100000000"],

        // coin inputs used to fund the fees (same as regular Txs)
        "coininputs": [],
        // optional, and is only allowed to be used for refunds
        "coinoutputs": [],

        // multi signature (verifiable using the given public keys),
        // (ed25519 by default, and only algo supported for now),
        // the order is always the same: first receiver, than sender.
        // As the receiver knows all info to define, it makes sense that sender
        // is the second signer always, as all the sender has to do is verify and agree.
        "signatures": [
            "433f5283a82fac28dacddeb98ed57aabb649a4aad2e7813af8a0009e0d035f625724dc7ef9ef39e75aef10fc77c4ed43fa0ce09f80c77d81ffd0c04ee7ca8c00"
            "b98ed57aabb649a4aad2e7813af8a0009e0d035f625724dc7ef9ef39e75aef10fc77c4ed43fa0ce09f80c77d81ffd0c04ee7ca8c00433f5283a82fac28dacdde"
        ]
    }
}
```

but it will have to be worked out further, before we can actually implement it with confidence.

**Comment by GlenDC** [comment]
All questions have been resolved, tfchain is now mostly feature complete for this story.

What has been done:
* all 3Bot transaction types are supported, in terms of signing, validation and encoding;
* 3Bot transactions are supported by the `tfchain` `transactionDB`, maintaining an internal DB of all 3Bot records;
* 3Bot records can be looked up using the REST API;
* 3Bot records can be looked up using the (tfchain) web explorer;
* 3Bots can be registered, updated and participate in name transfers using the CLI (which internally uses the REST API support to fulfil these requests);
* New primitive code has been unit tested and validated, more "high level" code still has to be unit tested;
* All new features have been validated manually on devnet, and seem to work;

Still to be done:
* add documentation about: new transaction types (encoding and signing of those transactions, 3bot feature and all concepts that go with it);
* resolve the double-registration/update (bad) bug (see #195);
* provide 3Bot support in the `rexplorer` binary;

**Comment by GlenDC** [comment]
Done:
* provide 3Bot support in the rexplorer binary;
* add documentation about: new transaction types (encoding and signing of those transactions, 3bot feature and all concepts that go with it);
* 3bot fees as miner fees;

Not done, but not important enough not to launch the first rc1 release of 1.1.2:
* resolve the double-registration/update (bad) bug (see #195);
